### PR TITLE
ARTEMIS-3178 Page Limitting (max messages and max bytes)

### PR DESCRIPTION
**Pull request by clebertsuconic**
I am adding three attributes to Address-settings:

* page-limit-bytes: Number of bytes. We will convert this metric into max number of pages internally by dividing max-bytes / page-size. It will allow a max based on an estimate.
* page-limit-messages: Number of messages
* page-full-message-policy: fail or drop

We will now allow paging, until these max values and then fail or drop messages.

Once these values are retracted, the address will remain full until a period where cleanup is kicked in by paging. So these values may have a certain delay on being applied, but they should always be cleared once cleanup happened.

### CODE REVIEW

**Comment by clebertsuconic** [comment]
I need to finish docs, and add testing for dropping.. which I should finish Monday morning.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 1]
```diff
@@ -225,13 +237,100 @@ public void applySetting(final AddressSettings addressSettings) {
       addressFullMessagePolicy = addressSettings.getAddressFullMessagePolicy();
 
       rejectThreshold = addressSettings.getMaxSizeBytesRejectThreshold();
+
+      pageFullMessagePolicy = addressSettings.getPageFullMessagePolicy();
+
+      pageLimitBytes = addressSettings.getPageLimitBytes();
+
+      pageLimitMessages = addressSettings.getPageLimitMessages();
```
Comment (gemmellr) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:245]: Should it validate you have actually set a policy if setting one or both of the others to enforcing values? That way you cant accidentally think you have a limit when you dont.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java [hunk 2]
```diff
@@ -179,11 +182,13 @@ public void close(PageSubscription cursor) {
...
 
    @Override
-   public void scheduleCleanup() {
+   public Future<Boolean> scheduleCleanup() {
+      final SimpleFutureImpl<Boolean> future = new SimpleFutureImpl<>();
       if (!cleanupEnabled || scheduledCleanup.intValue() > 2) {
          // Scheduled cleanup was already scheduled before.. never mind!
          // or we have cleanup disabled
-         return;
+         pagingStore.execute(() -> future.set(true));
```
Comment (tabish121) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java:190]: Is there a reason why you need to fire off a task into the executor just to return a completed future?  Seems simpler to just return CompletableFuture.completedFuture or just complete this one and return it directly.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 3]
```diff
@@ -1029,6 +1131,25 @@ public boolean page(Message message,
          lock.readLock().unlock();
       }
 
+      if (pageFull && pageFullMessagePolicy != null) {
```
Comment (gemmellr) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1134]: Related to earlier comment, if it validates you have a policy, can this then just simply check "if (pageFull)" ?

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd [hunk 4]
```diff
@@ -4062,6 +4079,20 @@
                </xsd:simpleType>
             </xsd:element>
 
+            <xsd:element name="page-full-policy" maxOccurs="1" minOccurs="0">
+               <xsd:annotation>
+                  <xsd:documentation>
+                     After entering page mode, a second limit will be set by page-limit-bytes, page-limit-messages. page-full-policy will configure what to do when that limit is reach.
```
Comment (gemmellr) [inline artemis-server/src/main/resources/schema/artemis-configuration.xsd:4085]: page-limit-bytes and/or page-limit-messages. The page-full-policy...

...when that limit is reach**ed.**

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java (hunk 2 above)
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java:190]: Yes... I wanted to make sure any previous call to cleanup was finished. Say you called scheduleCleanup a couple of times.. then you call scheduleCleanup again...


 I don't need to schedule another one as I already have a call scheduled, but I want to return a Future that will be released after these calls have finished.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java:190]: I am amending the comment before future.set call as:

         // Scheduled cleanup was already scheduled before.
         // On that case just flush the executor returning the future.set(true) 
         // after any previous scheduled cleanup is finished.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java:191]: marking it as resolved.. please unresolved it if you find any issues with my comments please?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/PagingStore.java [hunk 5]
```diff
@@ -60,6 +62,19 @@ public interface PagingStore extends ActiveMQComponent, RefCountMessageListener
 
    AddressFullMessagePolicy getAddressFullMessagePolicy();
 
+   PageFullMessagePolicy getPageFullMessagePolicy();
+
+   Long getPageLimitMessages();
```
[the remark this answers is not in the document]
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/PagingStore.java:67]: I was going against the -1 usage here. and just keep it as null.

I can edit the default on the XSD to state null

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd [hunk 6]
```diff
@@ -4000,6 +4000,23 @@
                </xsd:annotation>
             </xsd:element>
 
+            <xsd:element name="page-limit-bytes" type="xsd:string" default="-1" maxOccurs="1" minOccurs="0">
+               <xsd:annotation>
+                  <xsd:documentation>
+                     After the address entering into page mode, this attribute will configure how many pages can be written into page before activating the page-full-policy.
```
Comment (tabish121) [inline artemis-server/src/main/resources/schema/artemis-configuration.xsd:4006]: This reads a bit funny, should it be "After the address enters into page mode...."  same for the below comment.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 7]
```diff
@@ -225,13 +237,100 @@ public void applySetting(final AddressSettings addressSettings) {
...
+      pageFullMessagePolicy = addressSettings.getPageFullMessagePolicy();
+
+      pageLimitBytes = addressSettings.getPageLimitBytes();
+
+      pageLimitMessages = addressSettings.getPageLimitMessages();
+
+      if (pageLimitBytes != null && pageSize > 0) {
+         estimatedMaxPages = pageLimitBytes / pageSize;
+         logger.debug("Address {} should not allow more than {} pages", storeName, estimatedMaxPages);
+      }
```
[the remark this answers is not in the document]
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:250]: I'm changing the constructor to make any -1 as null.


I won't just change the File parser as the same could be set through properties.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 1 above)
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java]: added validation

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 3 above)
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1153]: Probably. It feels wrong not doing the redundant check though! :)

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java]: My OCD wants to keep it, but I will remove it :)

**Comment by clebertsuconic** [comment]
all the comments are resolved.. tests are 100%... I'm merging this...


if anyone sees anything else wrong I will send a new PR.. (or feel free to send any adjustments yourself of course).

**Comment by clebertsuconic** [comment]
thanks for the reviews!

**Pull request merged**